nexus/blueprint-execution/src/dns.rs

```diff
@@ -0,0 +1,744 @@
+    // `OmicronZoneType` variant instead.  Although that information is present,
+    // it's irritatingly non-trivial to do right now because SocketAddrs are
+    // represented as strings, so we'd need to parse all of them and handle all
+    // the errors, even though they should never happen.
+    // See oxidecomputer/omicron#4988.
+    for (_, omicron_zone) in blueprint.all_omicron_zones() {
+        if !blueprint.zones_in_service.contains(&omicron_zone.id) {
+            continue;
+        }
+
+        let (service_name, port) = match omicron_zone.zone_type {
+            OmicronZoneType::BoundaryNtp { .. } => {
+                (ServiceName::BoundaryNtp, NTP_PORT)
+            }
+            OmicronZoneType::InternalNtp { .. } => {
+                (ServiceName::InternalNtp, NTP_PORT)
+            }
+            OmicronZoneType::Clickhouse { .. } => {
+                (ServiceName::Clickhouse, CLICKHOUSE_PORT)
+            }
+            OmicronZoneType::ClickhouseKeeper { .. } => {
+                (ServiceName::ClickhouseKeeper, CLICKHOUSE_KEEPER_PORT)
+            }
+            OmicronZoneType::CockroachDb { .. } => {
+                (ServiceName::Cockroach, COCKROACH_PORT)
+            }
+            OmicronZoneType::Nexus { .. } => {
+                (ServiceName::Nexus, NEXUS_INTERNAL_PORT)
+            }
+            OmicronZoneType::Crucible { .. } => {
+                (ServiceName::Crucible(omicron_zone.id), CRUCIBLE_PORT)
+            }
+            OmicronZoneType::CruciblePantry { .. } => {
+                (ServiceName::CruciblePantry, CRUCIBLE_PANTRY_PORT)
+            }
+            OmicronZoneType::Oximeter { .. } => {
+                (ServiceName::Oximeter, OXIMETER_PORT)
+            }
+            OmicronZoneType::ExternalDns { .. } => {
+                (ServiceName::ExternalDns, DNS_HTTP_PORT)
+            }
+            OmicronZoneType::InternalDns { .. } => {
+                (ServiceName::InternalDns, DNS_HTTP_PORT)
+            }
+        };
+
+        // This unwrap is safe because this function only fails if we provide
+        // the same zone id twice, which should not be possible here.
+        dns_builder
+            .host_zone_with_one_backend(
+                omicron_zone.id,
+                omicron_zone.underlay_address,
+                service_name,
+                port,
+            )
+            .unwrap();
+    }
+
+    let scrimlets = sleds_by_id.values().filter(|sled| sled.is_scrimlet);
+    for scrimlet in scrimlets {
+        let sled_subnet = scrimlet.subnet();
+        let switch_zone_ip = get_switch_zone_address(sled_subnet);
+        // unwrap(): see above.
+        dns_builder
+            .host_zone_switch(
+                scrimlet.id,
+                switch_zone_ip,
+                DENDRITE_PORT,
+                MGS_PORT,
+                MGD_PORT,
+            )
+            .unwrap();
+    }
+
+    dns_builder.generation(blueprint.internal_dns_version.next());
```

Comment:
This threw me a bit, because it _looks_ like the returned DNS config has always has a generation bump, which means it will be marching up every time this RPW runs. But making sure I understand: that doesn't happen because the config we return here is compared against the existing config using `DnsDiff`, which does _not_ look at the generation number, only the contents, right? That seems a little implicit, but I don't have a great suggestion off the top of my head for making it clearer. We could wait to bump the generation until after the comparison, maybe? That might be a little more obvious?

Reply:
I see what you mean.  I think it would be an improvement if instead of `DnsConfigParams`, we passed around the equivalent of `dns_config_params.zones[0].records` -- i.e., just the _records_ that are in the internal DNS zone.  Then we'd assemble that into a final `DnsConfigParams` at the last minute, before writing it to the database.  Then it would be clearer what's being generated and diff'd, etc.  But right now there's no type for that, just `HashMap<String, Vec<DnsRecord>>`.  I think at minimum we'd want a newtype for this, and probably a richer one, too.  `DnsConfigBuilder` maybe should produce one of those instead of `DnsConfigParams`.  Anyway I think this would be better but I'd like to _not_ do it right now.

I have at least added a few comments to explain this though.

(I inferred your suggestion to "wait to bump the generation until after the comparison" to mean "return something with the same generation as what's there, then do the comparison, then mutate it if we find we need to bump it".  I don't love that.  It seems misleading in a different way.)